### PR TITLE
Bubble sort algorithm

### PR DESCRIPTION
**Pull request by alvarorf**
In this branch, we implement two versions of the bubble sort algorithm: 1) bubble_sort() which takes an array only as a parameter and 2) bubble_sort_by() which can also take a block as a parameter.

### CODE REVIEW

File: bubble_sort.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+def bubble_sort(uns_array)
+  return 'Wrong type of argument. It should be an array!' unless uns_array.is_a? Array
+
+  times_elts_were_swapped = 0
+  number_of_iterations_through_array = 0
+  uns_array.each do |_elt|
+    i = 0
+    while i < uns_array.length - number_of_iterations_through_array
+      spaceship_operator_result = uns_array[i] <=> uns_array[i + 1]
```
Comment (kitpao) [inline bubble_sort.rb:16]: - [ ] **OPTIONAL** It is nice that you use other operators aside from the mainstream ones, but the spaceship operation is more expensive in terms of efficiency than if you used a normal > (greater than) and you are not even using the case when it is == or <, so you don't need it

File: bubble_sort.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+# If swap count is greater than 0, repeat this process
+
+def bubble_sort(uns_array)
+  return 'Wrong type of argument. It should be an array!' unless uns_array.is_a? Array
+
+  times_elts_were_swapped = 0
+  number_of_iterations_through_array = 0
+  uns_array.each do |_elt|
+    i = 0
+    while i < uns_array.length - number_of_iterations_through_array
```
Comment (kitpao) [inline bubble_sort.rb:15]: - :clap: nice way of making your code efficient, avoiding to go through the whole array again if you know the last ones are already sorted. You could even skip more steps, if you want to know more, check at the end of this article the final code: https://hackernoon.com/bubble-sort-algorithm-how-to-develop-your-computational-thinking-5j3n30o2

File: bubble_sort.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    arr.each do |_el|
+      while i < arr.length - (i + 1)
+        comparison_result = yield(arr[i], arr[i + 1])
+        arr[i], arr [i + 1] = arr[i + 1], arr[i] if comparison_result.positive?
+        times_elts_were_swapped += 1 if comparison_result.positive?
+        i += 1
+      end
+      break if times_elts_were_swapped.zero?
+    end
+  end
```
Comment (kitpao) [inline bubble_sort.rb:44]: - [ ] It is a requirement in the tests that Microverse gives us, that this method should return an error when a block is not given. Your code returns nothing when this happens. Tests message: (Failure/Error: expect { bubble_sort_by(unsorted) }.to raise_error(LocalJumpError)). **Hint:** You can either Raise a LocalJumpError manually, or take out your if block_given and let the automatic error to be raised.

File: bubble_sort.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+puts '====== Testing bubble_sort ========'
+puts bubble_sort([6, 5, 4, 3, 2, 1])
+puts '========'
+puts bubble_sort([3, 5, 4, 0, 3, 2, 1])
+puts '========'
+puts bubble_sort([1, 2, 3, 4, 5, 6])
+puts '========'
+puts bubble_sort(%w[m c f z a k])
+puts '========'
+puts bubble_sort('random string')
```
Comment (kitpao) [inline bubble_sort.rb:59]: - [ ] Please, delete this code as it is considered debugging code. (In future projects you will learn how these tests are done automatically)
image

**Review by kitpao — CHANGES_REQUESTED** [review]
# 👋 Hello @alvarorf and @Blaise-Shyaka 

:star2: You did nicely in your logic, I see you tried to optimize your code by skipping redundant cases :+1: 
You are missing some details, please check below and in the inline comments :arrow_down: 

**_Note:_** All the **[OPTIONAL]** are **not required** in order to approve your project but are personal suggestions to improve your coding experience

- [ ] Your linters are not fixed yet, please correct them until you get a message similar to this one:
image

- [ ] **OPTIONAL** In your README you can be more specific in the setup, remember not everyone who sees your project is a programmer, so you can specify something like "in your terminal type` git clone git@github.com:alvarorf/microverse_bubble_sort.git` then cd to the folder and type `ruby bubble_sort.rb`"

- [ ] Your commits are descriptive, but for next time try to be consistent and write your commits in the imperative tense, as stated in [the article pointed in pathwright](https://www.freecodecamp.org/news/level-you-up-to-awesome-commit-messages-a85558cb90e8/)

- [ ] In both methods, you forgot to tell your program what to do if the input is already sorted, `p bubble_sort([1,2,3,4])` returns nil, and `p bubble_sort_by(%w[hi hey hello]) { |left, right| left.length - right.length }` also returns nil

### ♻️ Status: Your project requires some additional changes ♻️
➡️ Please request another review once you are ready 📝 
___
##### 👉 If you have any doubts you can contact me in slack as @kitpao
##### 🌟  If you have feedback about my reviews I would appreciate if you leave me a recommendation in LinkedIn 💌 or let me know in Slack if you see improvements to be done :smile: 

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

File: bubble_sort.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    return uns_array if times_elts_were_swapped.zero?
+  end
+end
+
+def bubble_sort_by(arr)
+  i = 0
+  times_elts_were_swapped = 0
+
+  arr.each do |_el|
+    while i < arr.length - (i + 1)
```
Comment (expjazz) [inline bubble_sort.rb:34]: - [ ] Remove the ( i + 1). If you leave the arr.length - 1 it works.

**Review by expjazz — CHANGES_REQUESTED** [review]
## Status: Changes Required :recycle: 

Congratulations on working on the changes requested. We are almost there :1st_place_medal: 

- [ ] You have one problem with the bubble_sort_by (check the inline comments below for more detail). It is not quite working yet.

image

In general, the bubble sort is slightly different. Since yours is working, just make this one small change so it passes the tests from Microverse.

### Please make the required changes and resubmit for a new code review once you are ready
Good Luck and Happy Coding :pray:

_P.S. If you have any questions about my review or need help, you can message me on slack (@ Expedito Andrade).:smile:_

**Review by tirthajyoti-ghosh — APPROVED** [review]
# Approved ✅

## You have done an incredible job making this project 👏🧨💪! Just perfect 💯! 

### You have met all the requirements of the project. Go ahead and merge it! Onto the next project 🚀 !

Happy coding 👨‍💻 

<hr>

**Tirthajyoti Ghosh**

[GitHub](https://github.com/tirthajyoti-ghosh) | [LinkedIn](https://www.linkedin.com/in/tirthajyoti-ghosh/) | [Twitter](https://twitter.com/terrific_ghosh)

**Pull request merged**